File: app/src/main/java/com/gabpawang/app/AppViewModel.kt
package com.gabpawang.app

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.gabpawang.app.data.WorkoutRepository
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import com.gabpawang.app.data.db.WorkoutSessionEntity

/**
 * ViewModel exposing DB-backed state (totalPushups, charStage, streak) and
 * social sign-in actions (Google, Kakao) as Compose-friendly StateFlows.
 */
class AppViewModel(application: Application) : AndroidViewModel(application) {
    private val repo: WorkoutRepository = (application as GabpaApplication).repository
    val totalPushups: StateFlow<Int> = repo.totalReps.stateIn(
        viewModelScope, SharingStarted.WhileSubscribed(5000), 0
    )

    val charStage: StateFlow<Int> = totalPushups.map { stageFor(it) }.stateIn(
        viewModelScope, SharingStarted.WhileSubscribed(5000), 1
    )

    val streak: StateFlow<Int> = repo.streakFlow().stateIn(
        viewModelScope, SharingStarted.WhileSubscribed(5000), 0
    )

    val oneRepMax: StateFlow<Int> = repo.allSessions.map { sessions ->
        sessions.flatMap { s ->
            s.setHistory.split(",").mapNotNull { it.trim().toIntOrNull() }
                .ifEmpty { listOf(s.totalReps) }
        }.maxOrNull() ?: 0
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), 0)

    /** Persists a finished workout result to the database. */
    fun saveWorkout(result: WorkoutResult, mode: String = "") {
        viewModelScope.launch {
            repo.saveSession(
                totalReps = result.total,
                sets = result.sets,
                durationSec = result.durationSec,
                history = result.history,
                mode = mode
            )
        }
    }

}

File: ios/GabpaWang/App/AppViewModel.swift
import Foundation
import Combine
import SwiftData

/// DB-backed 상태(누적/단계/연속/1RM)를 노출. `AppViewModel.kt` 포팅.
///
/// SwiftData가 KVO 없이 변경 알림을 잘 안 흘리기 때문에, 명시적으로
/// `refresh()`를 호출하는 흐름을 사용한다.
@MainActor
final class AppViewModel: ObservableObject {
    private let repository: WorkoutRepository

    @Published private(set) var totalPushups: Int = 0
    @Published private(set) var oneRepMax: Int = 0
    @Published private(set) var streak: Int = 0

    var charStage: Int { stageFor(total: totalPushups) }

    init(repository: WorkoutRepository) {
        self.repository = repository
        refresh()
    }

    func refresh() {
        totalPushups = repository.totalReps()
        oneRepMax = repository.oneRepMax()
        streak = repository.currentStreak()
    }

    /// 운동 종료 시 호출.
    func saveWorkout(_ result: WorkoutResult, mode: String = "") async {
        _ = await repository.saveSession(
            totalReps: result.total,
            sets: result.sets,
            durationSec: result.durationSec,
            history: result.history,
            mode: mode
        )
        refresh()
    }

    /// `RecordView` 등에서 직접 세션 목록을 조회할 때 사용.
    func allSessions() -> [WorkoutSessionEntity] {
        repository.fetchAllSessions()
    }
}
